text/0000-adt-kinds.md

```diff
@@ -0,0 +1,181 @@
+- Feature Name: clarified_adt_kinds
+- Start Date: 2016-02-07
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provide a simple model describing three kinds of structs and variants and their relationships.  
+Provide a way to match on structs/variants in patterns regardless of their kind (`S{..}`).  
+Permit tuple structs and tuple variants with zero fields (`TS()`).
+
+# Motivation
+[motivation]: #motivation
+
+There's some mental model lying under the current implementation of ADTs, but it is not written
+out explicitly and not implemented completely consistently.
+Writing this model out helps to identify its missing parts.
+Some of this missing parts turn out to be practically useful.
+This RFC can also serve as a piece of documentation.
+
+# Detailed design
+[design]: #detailed-design
+
+The text below mostly talks about structures, but almost everything is equally applicable to
+variants.
+
+## Braced structs
+
+Braced structs are declared with braces (unsurprisingly).
+
+```
+struct S {
+    field1: Type1,
+    field2: Type2,
+    field3: Type3,
+}
+```
+
+Braced structs are the basic struct kind, other kinds are built on top of them.
+Braced structs have 0 or more user-named fields and are defined only in type namespace.
+
+Braced structs can be used in struct expressions `S{field1: expr, field2: expr}`, including
+functional record update (FRU) `S{field1: expr, ..s}`/`S{..s}` and with struct patterns
+`S{field1: pat, field2: pat}`/`S{field1: pat, ..}`/`S{..}`.
+In all cases the path `S` of the expression or pattern is looked up in the type namespace (so these
+expressions/patterns can be used with type aliases).
+Fields of a braced struct can be accessed with dot syntax `s.field1`.
+
+Note: struct *variants* are currently defined in the value namespace in addition to type namespace,
+ there are no particular reasons for this and this is probably temporary.
```

Comment:
I guess this is because regular variants are in both namespaces and braced variants are a mistake. However, if we one day get named fields for functions, then I guess we could use struct names as ctor functions, so it might be useful. Too bad that ship has sailed for regular structs.